File: android/app/src/main/java/com/polemicyst/android/data/api/ApiError.kt
package com.polemicyst.android.data.api

import com.squareup.moshi.JsonClass

/**
 * Structured error body returned by the API on 403 and other error responses.
 * Different error codes include different subsets of fields.
 */
@JsonClass(generateAdapter = true)
data class ApiError(
    val code: String? = null,
    val message: String? = null,
    val error: String? = null,
    val allowedProviders: List<String>? = null,
    val plan: String? = null,
    val limit: Int? = null,
    val usage: Int? = null,
) {
    val displayMessage: String
        get() = message ?: error ?: "An unknown error occurred"

    val isQuotaExceeded: Boolean get() = code == CODE_QUOTA_EXCEEDED
    val isPlanRestricted: Boolean get() = code == CODE_PLAN_RESTRICTED

    companion object {
        const val CODE_QUOTA_EXCEEDED = "QUOTA_EXCEEDED"
        const val CODE_PLAN_RESTRICTED = "PLAN_RESTRICTED"
    }
}

/**
 * Exception carrying the parsed [ApiError] and HTTP status code.
 */
class ApiException(
    val statusCode: Int,
    val apiError: ApiError,
) : Exception(apiError.displayMessage)

File: src/lib/api-error.ts
export interface ApiQuotaError {
  error: string;
  code: 'QUOTA_EXCEEDED' | 'PLAN_RESTRICTED';
  limit?: number;
  usage?: number;
  allowedProviders?: string[];
}

export function isQuotaErrorCode(code: string): code is ApiQuotaError['code'] {
  return code === 'QUOTA_EXCEEDED' || code === 'PLAN_RESTRICTED';
}

export async function parseApiError(res: Response): Promise<ApiQuotaError | null> {
  if (res.status !== 403) return null;
  try {
    const body = await res.json();
    if (body?.code && isQuotaErrorCode(body.code)) {
      return body as ApiQuotaError;
    }
  } catch {
    // response body isn't JSON or is malformed
  }
  return null;
}
